src/test_memcachedsolution.cpp

```diff
@@ -661,50 +860,71 @@ TYPED_TEST(SimpleMemcachedSolutionTest, AddKillSetSetDataContentionSet)
     EXPECT_EQ(Store::Status::OK, rc);
     EXPECT_EQ(data_out, data_in);
 
-    TypeParam::fail_memcached_instance(this->_memcached_instances.back());
+    TypeParam::trigger_failure(this);
 
-    data_in = "SimpleMemcachedSolutionTest.AddKillSetSetDataContentionSet_New1";
+    data_in = "MemcachedSolutionFailureTest.AddKillSetSetDataContentionSet_New1";
     rc = this->set_data(data_in, cas);
     EXPECT_TRUE((rc == Store::Status::DATA_CONTENTION) ||
                 (rc == Store::Status::OK));
 
-    if (rc == Store::Status::DATA_CONTENTION)
+    cas_loop_count = 0;
+    while ((rc == Store::Status::DATA_CONTENTION) && (cas_loop_count < 5))
     {
       rc = this->get_data(data_out, cas);
       EXPECT_EQ(Store::Status::OK, rc);
+      EXPECT_NE(data_out, data_in);
 
       rc = this->set_data(data_in, cas);
-      EXPECT_EQ(Store::Status::OK, rc);
+      EXPECT_TRUE((rc == Store::Status::DATA_CONTENTION) ||
+                  (rc == Store::Status::OK));
     }
-    else if (rc == Store::Status::OK)
-    {
-      std::string failed_data_in = "FAIL";
-      rc = this->set_data(data_in, cas);
-      EXPECT_EQ(Store::Status::DATA_CONTENTION, rc);
+    EXPECT_LT(cas_loop_count, 5);
+
+    std::string failed_data_in = "FAIL";
+    rc = this->set_data(failed_data_in, cas);
+    EXPECT_EQ(Store::Status::DATA_CONTENTION, rc);
+
+    rc = this->get_data(data_out, cas);
+    EXPECT_EQ(Store::Status::OK, rc);
+    EXPECT_EQ(data_out, data_in);
+
+    data_in = "MemcachedSolutionFailureTest.AddKillSetSetDataContentionSet_New2";
+    rc = this->set_data(data_in, cas);
+    EXPECT_TRUE((rc == Store::Status::DATA_CONTENTION) ||
+                (rc == Store::Status::OK));
 
+    cas_loop_count = 0;
+    while ((rc == Store::Status::DATA_CONTENTION) && (cas_loop_count < 5))
+    {
       rc = this->get_data(data_out, cas);
       EXPECT_EQ(Store::Status::OK, rc);
-      EXPECT_EQ(data_out, data_in);
+      EXPECT_NE(data_out, data_in);
 
-      data_in = "SimpleMemcachedSolutionTest.AddKillSetSetDataContentionSet_New2";
       rc = this->set_data(data_in, cas);
       EXPECT_EQ(Store::Status::OK, rc);
-
-      rc = this->get_data(data_out, cas);
-      EXPECT_EQ(Store::Status::OK, rc);
-      EXPECT_EQ(data_out, data_in);
     }
+    EXPECT_LT(cas_loop_count, 5);
 
-    TypeParam::restore_memcached_instance(this->_memcached_instances.back());
+    rc = this->get_data(data_out, cas);
+    EXPECT_EQ(Store::Status::OK, rc);
+    EXPECT_EQ(data_out, data_in);
+
+    TypeParam::fix_failure(this);
+
+    // Bounce the store to prevent failures in this iteration from affecting
+    // the next one.
+    delete this->_store; this->_store = NULL;
+    this->_store = new TopologyNeutralMemcachedStore("astaire.local",
+                                                     this->_resolver);
```

Comment:
We need it here because the testcase contains a big loop that performs the same test on multiple different keys, so that it is quite likely that the for at least one of the iterations the memcached we kill is the primary for the key. This isn't needed for the other testscases, as the fixture deletes the store at the end of each testcase. I will beef up the comment.